sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -678,10 +684,11 @@ object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
           falseValue = u.withNewChildren(Array(falseValue)))
 
       case u @ UnaryExpression(c @ CaseWhen(branches, elseValue))
-          if supportedUnaryExpression(u) && atMostOneUnfoldable(branches.map(_._2) ++ elseValue) =>
+          if supportedUnaryExpression(u) && atMostOneUnfoldable(branches.map(_._2) :+
+            elseValue.getOrElse(Literal(null, c.dataType))) =>
```

Comment:
If the CaseWhen has only one branch and has no else branch, we can still pushdown?

Reply:
Yes. It can push down more filters. For example:
```scala
spark.sql("CREATE TABLE t1(a int, b int) using parquet")
spark.sql("SELECT * FROM t1 WHERE CASE WHEN a > 10 THEN b END = 20").explain
```
Pushdown:
```
== Physical Plan ==
*(1) Filter (((isnotnull(a#0) AND isnotnull(b#1)) AND (a#0 > 10)) AND (b#1 = 20))
+- *(1) ColumnarToRow
   +- FileScan parquet default.t1[a#0,b#1] Batched: true, DataFilters: [isnotnull(a#0), isnotnull(b#1), (a#0 > 10), (b#1 = 20)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/yumwang/spark/SPARK-31890/spark-warehouse/org.apache.spark..., PartitionFilters: [], PushedFilters: [IsNotNull(a), IsNotNull(b), GreaterThan(a,10), EqualTo(b,20)], ReadSchema: struct<a:int,b:int>
```
Do not pushdown:
```
== Physical Plan ==
*(1) Filter (CASE WHEN (a#0 > 10) THEN b#1 END = 20)
+- *(1) ColumnarToRow
   +- FileScan parquet default.t1[a#0,b#1] Batched: true, DataFilters: [(CASE WHEN (a#0 > 10) THEN b#1 END = 20)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/yumwang/spark/SPARK-31890/spark-warehouse/org.apache.spark..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<a:int,b:int>
```

Reply:
Got it. Then we should probably update `atMostOneUnfoldable`
```
  private def atMostOneUnfoldable(exprs: Seq[Expression]): Boolean = {
    if (exprs.length < 2) return true
    val (foldables, others) = exprs.partition(_.foldable)
    foldables.nonEmpty && others.length < 2
  }
```

Reply:
Isn't this equivalent to just removing the `foldables.nonEmpty` condition?

Reply:
Make sense.